File: dereplication_pilot/amr_plasmid_within_between.py
#!/usr/bin/env python3
"""AMR-carrying plasmid contig clusters: within-participant vs. whole-cohort
(same two-level logic as amr_within_between.py, but for Tier-1 contig
clusters that are BOTH geNomad-plasmid-classified AND carry a real RGI AMR
hit -- the specific "AMR plasmid dynamics, within vs. between" question).
"""
import csv
from collections import defaultdict

PILOT_DIR = "/maps/projects/hansen_ol-AUDIT/scratch/NILU/metagenomes/assembly_annotation_wf/dereplication_pilot"


def main():
    # 1. First-level AMR-carrying plasmid clusters, from the already-computed join
    amr_plasmid_owner = {}  # first-level cluster_id -> participant
    with open(f"{PILOT_DIR}/plasmid_amr_join.tsv") as fh:
        reader = csv.DictReader(fh, delimiter="\t")
        for row in reader:
            if row["is_plasmid"] == "True" and row["has_amr"] == "True":
                amr_plasmid_owner[row["cluster_id"]] = row["participant"]

    within_total = len(amr_plasmid_owner)

    # 2. Second-level (cross-participant) contig clustering
    second_level_of = {}
    with open(f"{PILOT_DIR}/cross_participant/clu_cluster.tsv") as fh:
        for line in fh:
            line = line.rstrip("\n")
            if not line:
                continue
            rep_id, member_id = line.split("\t")
            second_level_of[member_id] = rep_id

    second_level_participants = defaultdict(set)
    unresolved = 0
    for cid, participant in amr_plasmid_owner.items():
        sl = second_level_of.get(cid)
        if sl is None:
            unresolved += 1
            continue
        second_level_participants[sl].add(participant)

    whole_cohort_distinct = len(second_level_participants)
    shared = sum(1 for s in second_level_participants.values() if len(s) > 1)
    specific = whole_cohort_distinct - shared

    lines = []
    lines.append("=== AMR-carrying plasmid contig clusters: within-participant vs. whole-cohort ===")
    lines.append(f"within-participant distinct AMR+plasmid clusters (summed across participants): {within_total}")
    lines.append(f"whole-cohort distinct AMR+plasmid clusters (cross-participant dedup too): {whole_cohort_distinct}")
    if within_total:
        lines.append(
            f"additional collapse from cross-participant sharing: "
            f"{within_total - whole_cohort_distinct} "
            f"({100.0*(within_total - whole_cohort_distinct)/within_total:.1f}%)"
        )
    lines.append("")
    if whole_cohort_distinct:
        lines.append(
            f"shared across >=2 participants: {shared} ({100.0*shared/whole_cohort_distinct:.1f}%)"
        )
        lines.append(
            f"participant-specific: {specific} ({100.0*specific/whole_cohort_distinct:.1f}%)"
        )
    if unresolved:
        lines.append(f"WARNING: {unresolved} first-level clusters had no second-level match")

    hist = defaultdict(int)
    for s in second_level_participants.values():
        hist[len(s)] += 1
    lines.append("")
    lines.append("histogram: n_participants_sharing_this_AMR+plasmid -> n_distinct_clusters")
    for k in sorted(hist):
        lines.append(f"  {k}: {hist[k]}")

    report = "\n".join(lines) + "\n"
    with open(f"{PILOT_DIR}/amr_plasmid_within_between.txt", "w") as fh:
        fh.write(report)
    print(report, end="")

    with open(f"{PILOT_DIR}/report_amr_plasmid_within_between.tsv", "w") as fh:
        fh.write("n_participants_sharing\tn_distinct_clusters\n")
        for k in sorted(hist):
            fh.write(f"{k}\t{hist[k]}\n")


if __name__ == "__main__":
    main()

File: dereplication_pilot/R/amr_plasmid_within_between.R
# R port of amr_plasmid_within_between.py -- same within-vs-between logic as
# amr_within_between.R, restricted to contig clusters that are both geNomad-plasmid-
# classified and AMR-carrying (uses R_plasmid_amr_join.tsv from plasmid_amr_join.R).
# Run from dereplication_pilot/.

source("R/utils.R")

if (TRUE) {  # always run main logic on source() or Rscript alike -- sys.nframe()==0 broke under nested source()
  plasmid_amr <- read_tsv(file.path(PILOT_DIR, "R_plasmid_amr_join.tsv"),
                            col_types = cols(.default = "c")) %>%
    mutate(is_plasmid = is_plasmid == "TRUE", has_amr = has_amr == "TRUE")

  amr_plasmid_owner <- plasmid_amr %>% filter(is_plasmid, has_amr) %>% select(participant, cluster_id)
  within_total <- nrow(amr_plasmid_owner)

  second_level <- read_tsv(file.path(PILOT_DIR, "cross_participant/clu_cluster.tsv"),
                            col_names = c("rep_id", "member_id"), col_types = "cc", progress = FALSE)

  resolved <- amr_plasmid_owner %>% inner_join(second_level, by = c("cluster_id" = "member_id"))
  unresolved <- nrow(amr_plasmid_owner) - nrow(resolved)

  per_cluster <- resolved %>%
    group_by(rep_id) %>%
    summarise(n_participants_sharing = n_distinct(participant), .groups = "drop")

  whole_cohort_distinct <- nrow(per_cluster)
  shared <- sum(per_cluster$n_participants_sharing > 1)

  cat(sprintf("within-participant distinct AMR+plasmid clusters: %d\n", within_total))
  cat(sprintf("whole-cohort distinct: %d\n", whole_cohort_distinct))
  cat(sprintf("shared across >=2 participants: %d (%.1f%%)\n",
              shared, 100 * shared / whole_cohort_distinct))
  cat(sprintf("max participants sharing one: %d\n", max(per_cluster$n_participants_sharing)))
  if (unresolved > 0) cat(sprintf("WARNING: %d unresolved\n", unresolved))

  hist_tbl <- per_cluster %>% count(n_participants_sharing, name = "n_distinct_clusters")
  write_tsv(hist_tbl, file.path(PILOT_DIR, "R_report_amr_plasmid_within_between.tsv"))
}
